api/src/wfl/module/sg.clj

```diff
@@ -20,45 +25,123 @@
-      (run! update-record! (batch/submit-workload! workload env workflow-wdl cromwellify-workflow-inputs cromwell-label
-                                                   default-options))
+          default-options (util/deep-merge
+                           (util/make-options env)
+                           {:final_workflow_outputs_dir output
+                            #_#_:use_relative_output_paths true})]
+      (run! update-record! (batch/submit-workload!
+                            workload env workflow-wdl
+                            cromwellify-workflow-inputs cromwell-label
+                            default-options))
       (jdbc/update! tx :workload {:started now} ["id = ?" id]))
     (workloads/load-workload-for-id tx id)))
 
-(defoverload workloads/create-workload! pipeline create-sg-workload!)
-(defoverload workloads/start-workload! pipeline start-sg-workload!)
-(defoverload workloads/update-workload! pipeline batch/update-workload!)
-(defoverload workloads/load-workload-impl pipeline batch/load-batch-workload-impl)
+(defn clio-bam-record
+  "Return `nil` or the single Clio record with metadata `bam`, or throw."
+  [bam]
+  (let [records (clio/query-bam bam)
+        n       (count records)]
+    (when (> n 1)
+      (throw (ex-info "More than 1 Clio BAM record"
+                      {:bam_record bam :count n})))
+    (first records)))
+
+(defn clio-cram-record
+  "Return the useful part of the Clio record for `input_cram` or throw."
+  [input_cram]
+  (let [records (clio/query-cram {:cram_path input_cram})
+        n       (count records)]
+    (when (not= 1 n)
+      (throw (ex-info "Expected 1 Clio record with cram_path"
+                      {:cram_path input_cram :count n})))
+    (-> records first (select-keys [:billing_project
+                                    :data_type
+                                    :document_status
+                                    :insert_size_metrics_path
+                                    :location
+                                    :notes
+                                    :project
+                                    :sample_alias
+                                    :version]))))
+
+(defn register-workflow-in-clio
+  "Ensure Clio knows the `output` files for `_workflow` of `pipeline`."
+  [output pipeline {:keys [inputs uuid] :as _workflow}]
+  (let [{:keys [base_file_name input_cram]} (util/parse-json inputs)
+        base_file_name (or base_file_name
+                           (-> input_cram util/leafname
+                               (util/unsuffix ".cram")))
+        parts [output pipeline uuid pipeline "execution" base_file_name]
+        path  (partial str (str/join "/" parts))
+        bam   {:bai_path                   (path ".bai")
+               :bam_path                   (path ".bam")
+               :insert_size_metrics_path   (path ".insert_size_metrics")}]
+    (if-let [bam-record (clio-bam-record (select-keys bam [:bam_path]))]
+      bam-record
```

Comment:
So this is where the "ping clio per workflow" happens? And we want to make this `clio-bam-record` "read" happen only once per workflow?

Reply:
Yes. Exactly.